Enable GeoJson support in Azure.Search.Documents

Mohit-Chakraborty opened the pull request.

Mohit-Chakraborty committed:
Enable GeoJson support in Azure.Search.Documents

Mohit-Chakraborty committed:
Update formatter for the spatial types
Create proxy types for the GeoJson objects
Update tests

Mohit-Chakraborty committed:
Merge branch 'master' into EnableGeoJsonSupport

Commit:
[TSI] Disable getEvents test as it is blocking net-core from deploying (#19695)

RezaJooyandeh committed:
[Communication] - Sms - Return pagable results instead of array and update identity swagger (#19692)

pakrym committed:
Update Azure.Core to 1.11.0 (#19702)

dpokluda committed:
Update changelog for beta2 release (#19701)

heaths commented on sdk/search/Azure.Search.Documents/src/Azure.Search.Documents.csproj:
```diff
@@ -41,4 +41,7 @@
     <PackageReference Include="System.Threading.Channels" />
   </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\..\..\core\Azure.Core.Experimental\src\Azure.Core.Experimental.csproj" />
```
This should be conditioned on some property we can set (or otherwise unset). See sdk\search\Azure.Search.Documents\Directory.Build.props for examples. This makes it easy to condition in (or out) as needed if a servicing release had to be done, or it took longer than expected for this experimental feature. We did this with `FieldBuilder` support for spatial types and JSON object serializer (or something like that).

Mohit-Chakraborty replied:
Thanks for the hint. I am removing the reference from the project file.

heaths commented on sdk/search/Azure.Search.Documents/src/Indexes/FieldBuilder.cs:
```diff
@@ -35,9 +33,7 @@ public class FieldBuilder
                     [typeof(bool)] = SearchFieldDataType.Boolean,
                     [typeof(DateTime)] = SearchFieldDataType.DateTimeOffset,
                     [typeof(DateTimeOffset)] = SearchFieldDataType.DateTimeOffset,
-#if EXPERIMENTAL_SPATIAL
```
If types are still in Azure.Core.Experimental, it seems all these preproc conditions throughout the PR shouldn't be removed yet.

Mohit-Chakraborty replied:
I was under the impression that since this code is now being enabled and made production ready, we don't want it behind a macro anymore. The macro was to disable the code itself, and not about whether the code depends on `Azure.Core.Experimental`,
The preview version of the `Search` package can depend on `Azure.Core.Experimental` with no issues. Or, so I have been told. :)

heaths commented on sdk/search/Azure.Search.Documents/src/Spatial/GeoLineStringProxy.cs:
```diff
@@ -0,0 +1,39 @@
+namespace Azure.Search.Documents
+{
+    /// <summary>
+    /// Proxy for a <see cref="Azure.Core.GeoJson.GeoLineString"/> class.
```
Are you adding proxies (or was this one file just detected as new, but was moved/edited (a lot)? If so, why add new proxies? The idea was to get rid of them once GeoJSON types were added. If this is still experimental, I agree for now this is the right thing to do, but since the `EXPERIMENTAL_SPATIAL` preproc conditions were removed, I'm confused as to what the goal is in this PR.

Mohit-Chakraborty replied:
Maybe I misunderstood the purpose of the proxy objects.
I thought they were present to allow unknown objects to be serialized into GeoJson object types wherever possible. Is that not the case?

Are you saying we can remove all the code related to proxy objects?

heaths commented on sdk/search/Azure.Search.Documents/src/Spatial/GeoObjectProxy.cs:
```diff
@@ -86,5 +87,35 @@ protected T GetPropertyValue<T>(ref PropertyInfo property, string name)
+            if (proxies is null)
+            {
+                IEnumerable<GeoPosition> _list = GetPropertyValue<IEnumerable<GeoPosition>>(ref property, name);
+                var list = new List<GeoPosition>(_list);
+
+                List<GeoPositionProxy> _proxies = new (list.Count);
```
Nit: would be nice to stick with a style: `var` + explicit constructor, or explicit type and just `new(...)`. These are just whitespaces apart.

annelo-msft committed:
Add ACR delete tag methods (#19680)

- Add delete tag methods to ContainerRepositoryClient
- Use Track 1 mgmt plane ACR client library to import images for delete tests
- Refactor client tests so that tests setting read/write attributes to false are marked NonParallelizable

Addresses tag delete in https://github.com/azure/azure-sdk-for-net/issues/19547

heaths commented on sdk/search/Azure.Search.Documents/tests/Serialization/SearchFilterTests.cs:
```diff
@@ -178,30 +174,29 @@ public void Polygons()
-            Assert.AreEqual(
-                "geo.intersects(Foo, geography'POLYGON((0 0,0 1,1 1,0 0))')",
-                SearchFilter.Create($"geo.intersects(Foo, {polygon})"));
+            //GeoPolygon polygon = new GeoPolygon(new[] { line });
+            //Assert.AreEqual(
+            //    "geo.intersects(Foo, geography'POLYGON((0 0,0 1,1 1,0 0))')",
+            //    SearchFilter.Create($"geo.intersects(Foo, {polygon})"));
```
Nit: could just clean up the code to remove these since it's all in source control.

heaths commented on sdk/search/Azure.Search.Documents/tests/Serialization/SearchFilterTests.cs:
```diff
@@ -178,30 +174,29 @@ public void Polygons()
             Assert.Throws<ArgumentException>(() => SearchFilter.Create(
-                $"{new GeoPolygon(new[] { line, line })}"));
+                $"{new GeoLineString(new[] { new GeoPosition(0, 0), new GeoPosition(0, 0), new GeoPosition(0, 0), new GeoPosition(1, 1) })}"));
+            //Assert.Throws<ArgumentException>(() => SearchFilter.Create(
```
Nit: could just clean up the code to remove these since it's all in source control.

heaths commented on sdk/search/Azure.Search.Documents/tests/Serialization/SearchFilterTests.cs:
```diff
@@ -178,30 +174,29 @@ public void Polygons()
         private static IEnumerable GetMicrosoftSpatialPointsData()
         {
-            GeographyPoint point = GeographyPoint.Create(3.0, 2.0);
+            GeoPoint point = new (2.0, 3.0);
```
+1. It bugged me so much (and tripped me up several times) that `GeographyPoint` reversed params from serialized representation and, IIRC, the `GeographyPosition` ctor params.

heaths commented on sdk/search/Azure.Search.Documents/tests/Spatial/SpatialProxyFactoryTests.cs:
```diff
@@ -13,125 +13,125 @@ public class SpatialProxyFactoryTests
-                .LineTo(0.0, 1.0)
-                .LineTo(0.0, 0.0)),
+            new TestCaseData(new GeoPosition(1.0, 2.0)),
+            //new TestCaseData(new GeoLineString(new GeoPosition[]
```
Nit: could just clean up the code to remove these since it's all in source control.

heaths commented on sdk/search/Azure.Search.Documents/tests/Utilities/TestExtensions.cs:
```diff
@@ -161,49 +156,40 @@ public static string GetName(this Random random, int length = 8)
+        ///// </summary>
+        ///// <param name="value">The <see cref="GeographyPoint"/> to convert.</param>
+        ///// <returns>A <see cref="SearchDocument"/> for the given <paramref name="value"/>.</returns>
+        //public static SearchDocument AsDocument(this GeographyPoint value)
```
Nit: could just clean up the code to remove these since it's all in source control.

But, this isn't used anymore? Interesting.

Mohit-Chakraborty replied:
I have uncommented and fixed the code. I did not see it being used anywhere.

heaths reviewed (changes requested):
Overall looks pretty good, but I'm confused whether this is still experimental or not - which seems it is, since we now reference the Azure.Core.Experimental project. So shouldn't the preproc conditions stay in?

pakrym committed:
Resolve Operation type conflict in the deviceupdate sample (#19709)

pakrym committed:
Prepare Azure.Core.Experimental 0.1.0-preview.11 release (#19710)

RezaJooyandeh committed:
[Communication] - PhoneNumbers - Switch to non-generic Operation for void long running operations (#19707)

chidozieononiwu committed:
Cleanup eng\scripts\Language-Settings.ps1 (#19651)

* Cleanup eng\scripts\Language-Settings.ps1

* Fix bug in Langaugesettings, due to projectPaths being either string or array

* Update eng/scripts/Language-Settings.ps1

Co-authored-by: Wes Haggard <weshaggard@users.noreply.github.com>

Co-authored-by: Wes Haggard <weshaggard@users.noreply.github.com>

jonathan-innis committed:
[KubernetesConfiguration] Release preview version for Extensions (#19576)

* Initial commit

* Update extension instance tests with identity

* Update .NET SDK info

* Update extension instance model and tests

HarveyLink committed:
Update for Track 1 Code Generation validation improvements (#19013)

* Update Track1CodeCheck.ps1

* Update storage test

* Update Track1CodeCheck.ps1

* Update Track1CodeCheck.ps1

* Update Track1CodeCheck.ps1

* Update double test

* Update Track1CodeCheck.ps1

* Update Track1CodeCheck.ps1

* Revert "Update storage test"

This reverts commit 34809dd68bdd2d14451b34b5f804906fec68f599.

* Revert "Update double test"

This reverts commit 388f998ead062106db713c2595dd736bef93e4af.

* Update Track1CodeCheck.ps1

* Update Track1CodeCheck.ps1

* Update Track1CodeCheck.ps1

* Storage test

* Update Track1CodeCheck.ps1

* Revert "Storage test"

This reverts commit e2c363664406705998fea8c44f57fd7f2233e9e1.

* Update Directory.Build.Mgmt.props

* Update Track1CodeCheck.ps1

* Revert "Revert "Storage test""

This reverts commit 6dfe94822952df7b77ce98d73f4df2b8b2af5975.

* Compatibility test

* Undo Test

abatishchev committed:
Updated CP DeviceUpdate to 1.0.1-beta.1 (#19654)

* Update Microsoft.Azure.Management.DeviceUpdate.csproj

* Update Microsoft.Azure.Management.DeviceUpdate.csproj

* Update AssemblyInfo.cs

RezaJooyandeh committed:
[Communication] - PhoneNumbers - Hide unused types and making operations testable (#19715)

RezaJooyandeh committed:
[Communication] - PhoneNumbers - Remove unused type (#19719)

azure-sdk committed:
Add ReleaseTrackingOnly option to Prepare-Release (#19718)

Co-authored-by: Wes Haggard <Wes.Haggard@microsoft.com>

JoshLove-msft committed:
Add new ACS events and update the constant values for (#19705)

AcsChatParticipantAddedToThread and AcsChatParticipantRemovedFromThread
as the values were changed in the swagger.
Note that the constant names are not being changed to avoid build breaks.

JoshLove-msft committed:
InitialOffsetOptions updates (#19700)

* InitialOffsetOptions updates

* Update tests

* fix

* Add docs

maririos committed:
[Translator] Enabling live tests pipeline (#19711)

* fixing live tests

* fix variable names

a-noyass committed:
[documenttranslation] Lazy loaded pageable for GetValues() and override WaitForCompletion implementation (#19704)

Commit:
[TSI]: Test Improvements (#19726)

heaths committed:
Resolve a few flaky Key Vault tests (#19612)

* Wait longer for deleted keys

* Fix net5.0 tests on OSX

* Add EventSource for CertificateOperation

Added more logging to better diagnose how #17718 is happening.

* Always get start time from stopwatch

...in case the event source is initially disabled, but enabled in the middle (elapsed time would be very long).

* Fix broken test on OSX

* More fixes for OSX

* Resolve PR feedback

* Re-enable some tests and increasing polling interval

* Ignore VerifyECDecoderPrime256v1Imported on OSX

Mohit-Chakraborty commented:
> Overall looks pretty good, but I'm confused whether this is still experimental or not - which seems it is, since we now reference the Azure.Core.Experimental project. So shouldn't the preproc conditions stay in?

Yes, this Preview release of Search depends on `Azure.Core.Experimental`.
This is not experimental code however in Search's upcoming preview.

Mohit-Chakraborty committed:
Remove explicit reference from project file

Mohit-Chakraborty commented:
I will revive this PR after fixing the merge issues.
